plugin/xtrabackup/plugin.go

```diff
@@ -0,0 +1,306 @@
+// The list is of the form "databasename1[.table_name1] databasename2[.table_name2]".
+// If this option is not specified, all databases containing MyISAM and InnoDB tables will be backed up.
+//
+// mysql_datadir:
+// This option specifies MySQL's datadir.
+//
+// mysql_xtrabackup:
+// This option specifies the absolute path to the `xtrabackup` tool.
+//
+// mysql_temp_targetdir:
+// This option specifies the absolute path to a temporary directory used by
+// the `xtrabackup` tool to backup the MySQL databases. It must be empty after
+// each run of the plugin.
+//
+// mysql_tar:
+// This option specifies the absolute path to the `tar` tool.
+//
+//
+// BACKUP DETAILS
+//
+// The `xtrabackup` plugin backs up all data in the data directory. If the `databases` option is specified
+// the plugin will only back up these databases.
+//
+// RESTORE DETAILS
+//
+// To restore, the `xtrabackup` plugin moves back the backed up data files to
+// the MySQL data directory. The MySQL data directory needs to be empty before
+// the restore operation.
+//
+// DEPENDENCIES
+//
+// This plugin relies on the `xtrabackup` and `tar` utilities. Please ensure
+// that they are present on the system that will be running the
+// backups + restores for MySQL.
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/starkandwayne/goutils/ansi"
+
+	. "github.com/starkandwayne/shield/plugin"
+)
+
+var (
+	DefaultTar           = "tar"
+	DefaultDataDir       = "/var/lib/mysql"
+	DefaultTempTargetDir = "/tmp/backups"
+	DefaultXtrabackup    = "/var/vcap/packages/shield-mysql/bin/xtrabackup"
+)
+
+func main() {
+	p := XtraBackupPlugin{
+		Name:    "MySQL XtraBackup Plugin",
+		Author:  "-",
+		Version: "0.0.1",
+		Features: PluginFeatures{
+			Target: "yes",
+			Store:  "no",
+		},
+	}
+
+	Run(p)
+}
+
+type XtraBackupPlugin PluginInfo
+
+type XtraBackupEndpoint struct {
+	Databases string
+	DataDir   string
+	User      string
+	Password  string
+	Bin       string
+	TargetDir string
+	Tar       string
+}
+
+func (p XtraBackupPlugin) Meta() PluginInfo {
+	return PluginInfo(p)
+}
+
+func (p XtraBackupPlugin) Validate(endpoint ShieldEndpoint) error {
+	var (
+		s    string
+		err  error
+		fail bool
+	)
+
+	s, err = endpoint.StringValue("mysql_user")
+	if err != nil {
+		ansi.Printf("@R{\u2717 mysql_user          %s}\n", err)
+		fail = true
+	} else {
+		ansi.Printf("@G{\u2713 mysql_user}          @C{%s}\n", s)
+	}
+
+	s, err = endpoint.StringValue("mysql_password")
+	if err != nil {
+		ansi.Printf("@R{\u2717 mysql_password      %s}\n", err)
+		fail = true
+	} else {
+		ansi.Printf("@G{\u2713 mysql_password}      @C{%s}\n", s)
+	}
+
+	s, err = endpoint.StringValueDefault("mysql_databases", "")
+	if err != nil {
+		ansi.Printf("@R{\u2717 mysql_databases  %s}\n", err)
+		fail = true
+	} else if s == "" {
+		ansi.Printf("@G{\u2713 mysql_databases}  no databases\n")
+	} else {
+		ansi.Printf("@G{\u2713 mysql_databases}  @C{%s}\n", s)
+	}
```

Comment:
Should we verify the additional parameters that this plugin introduces, like `mysql_datadir`, `mysql_temp_targetdir`, `mysql_xtrabackup`, and `mysql_tar`?